File: CSharp/ByRound/1744/1744_E1_Divisible_Numbers_easy_version.cs
// Problem: CF 1744 E1 - Divisible Numbers (easy version)
// https://codeforces.com/contest/1744/problem/E1

﻿/*
 * Problem: Find a pair (x, y) such that:
 *   - a < x <= c
 *   - b < y <= d
 *   - x * y is divisible by a * b
 *
 * Algorithm:
 *   1. Generate all divisors of a and b.
 *   2. For each pair of divisors (d1, d2) of a and b respectively:
 *       - Compute x = d1 * (a / d1) * k1, y = d2 * (b / d2) * k2 (where k1, k2 are chosen so that x <= c, y <= d)
 *       - Check if there exist valid x and y within the range.
 *   3. Return first valid pair found or -1 -1 if none exists.
 *
 * Time Complexity: O(sqrt(a) * sqrt(b) + sqrt(a) * sqrt(b)) = O(sqrt(a)*sqrt(b))
 * Space Complexity: O(sqrt(a) + sqrt(b)) for storing divisors
 */
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;
namespace Compete2 {
partial class MainClass {
  public static void Main(string[] args) {
    int fyt = int.Parse(Console.ReadLine());
    List<string> output = new List<string>();
    for (int z = 0; z < fyt; z++) {
      var a =
          Console.ReadLine().Split(' ').Select(qw => long.Parse(qw)).ToList();
      List<long> o = new List<long>();
      List<long> a1 = new List<long>() { 1, a[0] }; // Divisors of a
      List<long> a2 = new List<long>() { 1, a[1] }; // Divisors of b
      // Find all divisors of a
      for (int i = 2; i < Math.Sqrt(a[0] + 1); i++) {
        if (a[0] % i == 0) {
          a1.Add(i);
          if (i * i != a[0])
            a1.Add(a[0] / i);
        }
      }
      // Find all divisors of b
      for (int i = 2; i < Math.Sqrt(a[1] + 1); i++) {
        if (a[1] % i == 0) {
          a2.Add(i);
          if (i * i != a[1])
            a2.Add(a[1] / i);
        }
      }
      a1 = a1.OrderBy(i => i).ToList(); // Sort divisors of a
      a2 = a2.OrderBy(i => i).ToList(); // Sort divisors of b
      long x = -1;
      long y = -1;
      bool bro = false;
      // Iterate through all combinations of divisors
      for (int i = 0; i < a1.Count; i++) {
        for (int j = 0; j < a2.Count; j++) {
          var x1 = a1[i] * a2[j];
          var y1 = (a[0] / a1[i]) * (a[1] / a2[j]);
          // Adjust x1 and y1 to make them within valid range
          if (x1 <= a[0]) {
            x1 = x1 * (a[0] / x1 + 1); // Scale up x1 to exceed a
          }
          if (y1 <= a[1]) {
            y1 = y1 * (a[1] / y1 + 1); // Scale up y1 to exceed b
          }
          // Check if x1 and y1 are within the given limits
          if (x1 <= a[2] && y1 <= a[3]) {
            x = x1;
            y = y1;
            bro = true;
            break;
          }
        }
        if (bro)
          break;
      }
      o.Add(x);
      o.Add(y);
      output.Add(string.Join(" ", o));
    }
    string on = string.Join("\n", output);
    Console.WriteLine(on);
  }
}
}


// https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1744/1744_E1_Divisible_Numbers_easy_version.cpp
// Problem: CF 1744 E1 - Divisible Numbers (easy version)
// https://codeforces.com/contest/1744/problem/E1

/*
Algorithm: 
- For each x in range (a, c], calculate the gcd of (a*b) and x.
- Use the gcd to compute k = (a*b) / gcd.
- Check if there exists a valid y in range (b, d] such that x*y is divisible by a*b.
- If such a y exists, output x and y, else continue searching.
- If no valid pair is found after checking all x, output -1 -1.

Time Complexity: O((c - a) * log(a*b)) where log(a*b) is for GCD computation.
Space Complexity: O(1) - only using a few variables.
*/

#include <algorithm>
#include <iostream>

typedef long long ll;
using namespace std;

int main() {
  int n;
  cin >> n;
  while (n--) {
    ll a, b, c, d, k, h = 1, m;
    cin >> a >> b >> c >> d;
    
    // Iterate through all possible values of x in range (a, c]
    for (ll x = a + 1; x <= c; x++) {
      m = __gcd(a * b, x); // Compute GCD of (a*b) and x
      k = a * b / m;       // Calculate k = (a*b) / gcd(a*b, x)
      
      // If b/k != d/k, there exists a valid y in range (b, d]
      // We compute y as (b/k + 1) * k which ensures that x*y is divisible by a*b
      if (b / k != d / k) {
        cout << x << " " << (b / k + 1) * k << endl;
        h = 0;
        break;
      }
    }
    
    // If no valid pair was found, output -1 -1
    if (h == 1) {
      cout << -1 << " " << -1 << endl;
    }
  }
}


// https://github.com/VaHiX/CodeForces/